# Scotland vs UK: Airport Passenger Recovery Analysis (2019–2023)

** How did Scotland’s airports recover from COVID compared to major UK airports?

# IMPORT REQUIRED LIBRARIES

In [569]:
# step 1: Import required libraries.

# -> CORE DATA ANALYSIS & CLEANING
import pandas as pd            # Pandas – load CSVs, clean data, merge datasets, groupby operations
import numpy as np             # NumPy – math operations, handling missing values

# -> DATA VISUALISATION (EDA + Dashboard)

import plotly.express as px    # Plotly Express – interactive charts (bar, line, heatmap, bubble)
import plotly.graph_objects as go   # Plotly Graph Objects – advanced custom visualizations


# -> STREAMLIT DASHBOARD FRAMEWORK
import streamlit as st         # Streamlit – build and deploy interactive web dashboards


# -> DATE & TIME FORMATTIN
from datetime import datetime  # Convert YYYYMM format to Year / Month columns

#  -> OPTIONAL FOR STYLING & WARNINGS
!pip install seaborn
import seaborn as sns  # Optional (EDA only)
import matplotlib.pyplot as plt  # Optional (EDA only)





# Data uploading (Table 09 Terminal and Transit Passengers)

In [570]:
# Step 2: Data Uploading (Table-09 (Terminal & Transit Passengers)

# step 2.1: Load 2019 Table-09 (Terminal & Transit Passengers)
df_2019_TT_Passengers=pd.read_csv("2019/Table 09 Terminal and Transit Passengers2019.csv")

# Step 2.2: Load 2020 Table-09 (Terminal & Transit Passengers)
df_2020_TT_Passengers=pd.read_csv("2020/Table 09 Terminal and Transit Passengers2020.csv")

# Step 2.3: Load 2021 Table-09 (Terminal & Transit Passengers)
df_2021_TT_Passengers=pd.read_csv("2021/Table_09_Terminal_and_Transit_Passengers2021.csv")

# Step 2.4: Load 2022 Table-09 (Terminal & Transit Passengers)
df_2022_TT_Passengers=pd.read_csv("2022/Table_09_Terminal_and_Transit_Passengers2022.csv")

# Step 2.5: Load 2023 Table-09 (Terminal & Transit Passengers)
df_2023_TT_Passengers=pd.read_csv("./2023/Table_09_Terminal_and_Transit_Passengers_2023.csv")

print("Data uploaded successfully!")


Data uploaded successfully!


# Data understanding (Table 09 Terminal and Transit Passengers 2019)

In [571]:
# Step 3: Data understanding.
# Step 3.1: Understanding the structure of data.
df_2019_TT_Passengers.shape

(52, 16)

****** Rows: 52

Columns: 16

Passenger Categories at Airports
__________________________________
There are two main types of passengers measured in airport statistics:
Type	Meaning	Example
Terminal Passengers	 -> Passengers who start or end their journey at that airport	Someone flying from Edinburgh to London, who checks in and leaves from Edinburgh
Transit Passengers ->	Passengers who pass through the airport but do not change flights (stay on same aircraft)	International service stopping to refuel or pick passengers but same flight continues


In [572]:
# Step 3.2: Preview the first five rows
df_2019_TT_Passengers.head()

rundate  this_period  last_period  airport_cluster  \
0  3/17/2020 10:40:30 AM         2019         2018                1   
1  3/17/2020 10:40:30 AM         2019         2018                1   
2  3/17/2020 10:40:30 AM         2019         2018                1   
3  3/17/2020 10:40:30 AM         2019         2018                1   
4  3/17/2020 10:40:30 AM         2019         2018                1   

   rpt_apt_grp_cd      rpt_apt_grp_name rpt_apt_name  total_pax_tp  \
0               1  London Area Airports      GATWICK      46576473   
1               1  London Area Airports     HEATHROW      80890031   
2               1  London Area Airports  LONDON CITY       5122271   
3               1  London Area Airports        LUTON      18216207   
4               1  London Area Airports     SOUTHEND       2035535   

   total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  term_pax_pc  \
0      46086089             1     46574786     46081327            1   
1      80124537             1     80886589     80100311            1   
2       4820403             6      5122271      4820403            6   
3      16769634             9     18213901     16766552            9   
4       1480139            38      2035535      1480139           38   

   tran_pax_tp  tran_pax_lp  tran_pax_pc  
0         1687         4762          -65  
1         3442        24226          -86  
2            0            0          NaN  
3         2306         3082          -25  
4            0            0          NaN

***** This helps verify that the data has loaded correctly .


In [573]:
# Step 3.3: Display all column names in the 2019 dataset
df_2019_TT_Passengers.columns

Index(['rundate', 'this_period', 'last_period', 'airport_cluster',
       'rpt_apt_grp_cd', 'rpt_apt_grp_name', 'rpt_apt_name', 'total_pax_tp',
       'total_pax_lp', 'total_pax_pc', 'term_pax_tp', 'term_pax_lp',
       'term_pax_pc', 'tran_pax_tp', 'tran_pax_lp', 'tran_pax_pc'],
      dtype='object')

# ****** For biginners -  easy to study *******
### Column Descriptions — 2019 Terminal & Transit Passengers Dataset

| Column Name | Meaning | Simple Explanation | Example |
---------------------------------------------

| **rundate** | Report Date | When the passenger data was generated or published | 3/17/2020 |

| **this_period** | Current Reporting Year | Year of the passenger statistics | 2019 |

| **last_period** | Previous Reporting Year | Used for year-to-year comparison | 2018 |

| **airport_cluster** | Airport Cluster ID | Numerical ID grouping similar airports | 1 = London area |

| **rpt_apt_grp_cd** | Airport Group Code | Code for region/group an airport belongs to | 1 |

| **rpt_apt_grp_name** | Airport Group Name | Geographic grouping of airports | London Area Airports |

| **rpt_apt_name** | Airport Name | Specific reporting airport | Heathrow, Gatwick |

| **total_pax_tp** | Total Passengers (This Period) | Total passengers for current year | 8,089,031 |

| **total_pax_lp** | Total Passengers (Last Period) | For previous year | 8,012,453 |

| **total_pax_pc** | Total Passenger % Change | Growth/Decline vs previous year | 0.95% |

| **term_pax_tp** | Terminal Passengers (This Period) | Passengers starting/ending journey | e.g., Boarding/Leaving |

| **term_pax_lp** | Terminal Passengers (Last Period) | Previous year terminal count | - |

| **term_pax_pc** | Terminal Passenger % Change | Terminal YoY growth | %

| **tran_pax_tp** | Transit Passengers (This Period) | Passengers who **do not change flight** (same aircraft continues) | e.g., Stopover passengers |

| **tran_pax_lp** | Transit Passengers (Last Period) | Previous year transit count | - |

| **tran_pax_pc** | Transit Passenger % Change | Transit YoY growth | %

CODE MEANING
------------------
Code	   Meaning	
rpt	 ->  Reporting	

apt	 ->  Airport

grp	 ->.  Group	

cd	 ->. Code

tp   ->	This Period	

lp	 -> Last Period	

pc	-> Percentage Change	Growth % between TP and LP

pax	-> Passengers	Number of passengers



In [574]:
# Step 3.4: Check basic info (data types, missing values, memory use)
df_2019_TT_Passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rundate           52 non-null     object 
 1   this_period       52 non-null     int64  
 2   last_period       52 non-null     int64  
 3   airport_cluster   52 non-null     int64  
 4   rpt_apt_grp_cd    52 non-null     int64  
 5   rpt_apt_grp_name  52 non-null     object 
 6   rpt_apt_name      52 non-null     object 
 7   total_pax_tp      52 non-null     int64  
 8   total_pax_lp      52 non-null     int64  
 9   total_pax_pc      51 non-null     float64
 10  term_pax_tp       52 non-null     int64  
 11  term_pax_lp       52 non-null     int64  
 12  term_pax_pc       51 non-null     float64
 13  tran_pax_tp       52 non-null     int64  
 14  tran_pax_lp       52 non-null     int64  
 15  tran_pax_pc       33 non-null     float64
dtypes: float64(3), int64(10), object(3)
memory usa

*****. Dataset Info Summary

- Total Rows: 52  
- Total Columns: 16  
- Memory Used: 6.6 KB  
- Data Types:
  - 10 integer columns
  - 3 float columns
  - 3 object (text) columns
- `rundate` is text → will convert to date format
- Some percentage columns have missing values (normal when transit passengers = 0)


In [575]:
# Step 3.5: Unique Airport Names in the Dataset
df_2019_TT_Passengers['rpt_apt_name'].unique()

array(['GATWICK', 'HEATHROW', 'LONDON CITY', 'LUTON', 'SOUTHEND',
       'STANSTED', 'ABERDEEN', 'BARRA', 'BELFAST CITY (GEORGE BEST)',
       'BELFAST INTERNATIONAL', 'BENBECULA', 'BIGGIN HILL', 'BIRMINGHAM',
       'BLACKPOOL', 'BOURNEMOUTH', 'BRISTOL', 'CAMPBELTOWN',
       'CARDIFF WALES', 'CITY OF DERRY (EGLINTON)', 'DONCASTER SHEFFIELD',
       'DUNDEE', 'EAST MIDLANDS INTERNATIONAL', 'EDINBURGH', 'EXETER',
       'GLASGOW', 'HUMBERSIDE', 'INVERNESS', 'ISLAY',
       'ISLES OF SCILLY (ST.MARYS)', 'KIRKWALL', 'LANDS END (ST JUST)',
       'LEEDS BRADFORD', 'LERWICK (TINGWALL)', 'LIVERPOOL (JOHN LENNON)',
       'LYDD', 'MANCHESTER', 'NEWCASTLE', 'NEWQUAY', 'NORWICH',
       'OXFORD (KIDLINGTON)', 'PRESTWICK', 'SCATSTA', 'SOUTHAMPTON',
       'STORNOWAY', 'SUMBURGH', 'TEESSIDE INTERNATIONAL AIRPORT', 'TIREE',
       'WICK JOHN O GROATS', 'ALDERNEY', 'GUERNSEY', 'ISLE OF MAN',
       'JERSEY'], dtype=object)

In [576]:
# Step 3.6: Display them nicely in a list
sorted(df_2019_TT_Passengers['rpt_apt_name'].unique())


['ABERDEEN',
 'ALDERNEY',
 'BARRA',
 'BELFAST CITY (GEORGE BEST)',
 'BELFAST INTERNATIONAL',
 'BENBECULA',
 'BIGGIN HILL',
 'BIRMINGHAM',
 'BLACKPOOL',
 'BOURNEMOUTH',
 'BRISTOL',
 'CAMPBELTOWN',
 'CARDIFF WALES',
 'CITY OF DERRY (EGLINTON)',
 'DONCASTER SHEFFIELD',
 'DUNDEE',
 'EAST MIDLANDS INTERNATIONAL',
 'EDINBURGH',
 'EXETER',
 'GATWICK',
 'GLASGOW',
 'GUERNSEY',
 'HEATHROW',
 'HUMBERSIDE',
 'INVERNESS',
 'ISLAY',
 'ISLE OF MAN',
 'ISLES OF SCILLY (ST.MARYS)',
 'JERSEY',
 'KIRKWALL',
 'LANDS END (ST JUST)',
 'LEEDS BRADFORD',
 'LERWICK (TINGWALL)',
 'LIVERPOOL (JOHN LENNON)',
 'LONDON CITY',
 'LUTON',
 'LYDD',
 'MANCHESTER',
 'NEWCASTLE',
 'NEWQUAY',
 'NORWICH',
 'OXFORD (KIDLINGTON)',
 'PRESTWICK',
 'SCATSTA',
 'SOUTHAMPTON',
 'SOUTHEND',
 'STANSTED',
 'STORNOWAY',
 'SUMBURGH',
 'TEESSIDE INTERNATIONAL AIRPORT',
 'TIREE',
 'WICK JOHN O GROATS']

*** Now airport names in alphabetical order, making it easier to read

In [577]:
# Step 3.7: Define airports of interest (Scotland + comparison)
airports_scotland=[ "ABERDEEN",
    "BARRA",
    "BENBECULA",
    "CAMPBELTOWN",
    "DUNDEE",
    "EDINBURGH",
    "GLASGOW",
    "INVERNESS",
    "ISLAY",
    "KIRKWALL",
    "LERWICK (TINGWALL)",
    "PRESTWICK",
    "SCATSTA",
    "STORNOWAY",
    "SUMBURGH",
    "TIREE",
    "WICK JOHN O GROATS"]
airports_compare = [
    "HEATHROW",
    "GATWICK",
    "MANCHESTER"
]

airports_of_interest = airports_scotland + airports_compare


In [578]:
# Step 3.8: Check descriptive statistics
df_2019_TT_Passengers.describe()

this_period  last_period  airport_cluster  rpt_apt_grp_cd  \
count           52           52               52              52   
mean         2,019        2,018                1               3   
std              0            0                0               1   
min          2,019        2,018                1               1   
25%          2,019        2,018                1               3   
50%          2,019        2,018                1               3   
75%          2,019        2,018                1               3   
max          2,019        2,018                2               4   

       total_pax_tp  total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  \
count            52            52            51           52           52   
mean      5,778,410     5,687,522            -1    5,775,602    5,684,343   
std      13,784,089    13,609,054            17   13,783,027   13,605,305   
min               0           100           -86            0          100   
25%          61,331        61,498            -3       61,174       61,232   
50%         828,992       759,186             0      828,902      742,747   
75%       4,768,056     4,833,760             6    4,766,747    4,833,742   
max      80,890,031    80,124,537            38   80,886,589   80,100,311   

       term_pax_pc  tran_pax_tp  tran_pax_lp  tran_pax_pc  
count           51           52           52           33  
mean            -1        2,808        3,179          185  
std             17        7,173        7,868          545  
min            -86            0            0          -98  
25%             -3            0            0          -40  
50%              0          184          184           -5  
75%              5        2,088        2,346           85  
max             38       33,976       37,827        2,600

In [579]:
# Step 3.9: Fix Number Formatting for Readability (Remove Scientific Notation)
pd.set_option('display.float_format', '{:,.0f}'.format)
df_2019_TT_Passengers.describe().T



count      mean        std   min    25%     50%       75%  \
this_period         52     2,019          0 2,019  2,019   2,019     2,019   
last_period         52     2,018          0 2,018  2,018   2,018     2,018   
airport_cluster     52         1          0     1      1       1         1   
rpt_apt_grp_cd      52         3          1     1      3       3         3   
total_pax_tp        52 5,778,410 13,784,089     0 61,331 828,992 4,768,056   
total_pax_lp        52 5,687,522 13,609,054   100 61,498 759,186 4,833,760   
total_pax_pc        51        -1         17   -86     -3       0         6   
term_pax_tp         52 5,775,602 13,783,027     0 61,174 828,902 4,766,747   
term_pax_lp         52 5,684,343 13,605,305   100 61,232 742,747 4,833,742   
term_pax_pc         51        -1         17   -86     -3       0         5   
tran_pax_tp         52     2,808      7,173     0      0     184     2,088   
tran_pax_lp         52     3,179      7,868     0      0     184     2,346   
tran_pax_pc         33       185        545   -98    -40      -5        85   

                       max  
this_period          2,019  
last_period          2,018  
airport_cluster          2  
rpt_apt_grp_cd           4  
total_pax_tp    80,890,031  
total_pax_lp    80,124,537  
total_pax_pc            38  
term_pax_tp     80,886,589  
term_pax_lp     80,100,311  
term_pax_pc             38  
tran_pax_tp         33,976  
tran_pax_lp         37,827  
tran_pax_pc          2,600

*** Total passengers analysis:

*Average passengers increased from 2018 → 2019

A few major airports dominate UK aviation.

Many smaller airports handled very little traffic.

Even before COVID, some airports were already struggling.



--------------------------------------------------------

Count	   ->           52	                                ( Number of airports)

Mean	->          5,778,410	                           (Average passengers per airport)

Std 	  ->          13,784,089	                           (Very large variation → airports differ hugely in scale)

Min	        ->          0	                              (Some airports had no reported passengers (or missing flights))

25% 	       ->      61,331	                             (25% of airports served fewer than ~61K passengers)

(50%)	        ->     828,992.    	                   (Typical airport served around 0.8 million passengers)

75% 	       -> 4,768,056	                       (Top 25% airports served more than 4.7 million passengers)

Max	           ->   80,890,031	                           (One major hub airport served ~81 million passengers)


# Data understanding (Table 09 Terminal and Transit Passengers 2020)

In [580]:
# Step 4: Data understanding
# Step 4.1: Understanding the structure of data.
df_2020_TT_Passengers.shape

(53, 16)

****** Rows: 53

Columns: 16

In [581]:
# Step 4.2: Preview the first five rows
df_2020_TT_Passengers.head()

rundate  this_period  last_period  airport_cluster  \
0  4/16/2021 5:12:43 PM         2020         2019                1   
1  4/16/2021 5:12:43 PM         2020         2019                1   
2  4/16/2021 5:12:43 PM         2020         2019                1   
3  4/16/2021 5:12:43 PM         2020         2019                1   
4  4/16/2021 5:12:43 PM         2020         2019                1   

   rpt_apt_grp_cd      rpt_apt_grp_name rpt_apt_name  total_pax_tp  \
0               1  London Area Airports      GATWICK      10173431   
1               1  London Area Airports     HEATHROW      22111326   
2               1  London Area Airports  LONDON CITY        908105   
3               1  London Area Airports        LUTON       5550821   
4               1  London Area Airports     SOUTHEND        401143   

   total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  term_pax_pc  \
0      46576473           -78     10171867     46574786          -78   
1      80890031           -73     22109550     80886589          -73   
2       5122271           -82       908105      5122271          -82   
3      18216207           -70      5550401     18213901          -70   
4       2035535           -80       401143      2035535          -80   

   tran_pax_tp  tran_pax_lp  tran_pax_pc  
0         1564         1687           -7  
1         1776         3442          -48  
2            0            0          NaN  
3          420         2306          -82  
4            0            0          NaN

***** Each row represents one airport's total passenger count for the reporting year.

2020 is a key year in aviation analysis because:

It represents the peak COVID impact

Passenger restrictions resulted in severe demand collapse

Travel volumes dropped by 70%–90% in many airports

In [582]:
# Step 4.3: Display all column names in the 2019 dataset
df_2020_TT_Passengers.columns

Index(['rundate', 'this_period', 'last_period', 'airport_cluster',
       'rpt_apt_grp_cd', 'rpt_apt_grp_name', 'rpt_apt_name', 'total_pax_tp',
       'total_pax_lp', 'total_pax_pc', 'term_pax_tp', 'term_pax_lp',
       'term_pax_pc', 'tran_pax_tp', 'tran_pax_lp', 'tran_pax_pc'],
      dtype='object')

In [583]:
 #Step 4.4: Check basic info (data types, missing values, memory use)
df_2020_TT_Passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rundate           53 non-null     object 
 1   this_period       53 non-null     int64  
 2   last_period       53 non-null     int64  
 3   airport_cluster   53 non-null     int64  
 4   rpt_apt_grp_cd    53 non-null     int64  
 5   rpt_apt_grp_name  53 non-null     object 
 6   rpt_apt_name      53 non-null     object 
 7   total_pax_tp      53 non-null     int64  
 8   total_pax_lp      53 non-null     int64  
 9   total_pax_pc      50 non-null     float64
 10  term_pax_tp       53 non-null     int64  
 11  term_pax_lp       53 non-null     int64  
 12  term_pax_pc       50 non-null     float64
 13  tran_pax_tp       53 non-null     int64  
 14  tran_pax_lp       53 non-null     int64  
 15  tran_pax_pc       31 non-null     float64
dtypes: float64(3), int64(10), object(3)
memory usa

In [584]:
# Step 4.5: Display  Airport names  nicely in a list
sorted(df_2020_TT_Passengers['rpt_apt_name'].unique())


['ABERDEEN',
 'ALDERNEY',
 'BARRA',
 'BELFAST CITY (GEORGE BEST)',
 'BELFAST INTERNATIONAL',
 'BENBECULA',
 'BIGGIN HILL',
 'BIRMINGHAM',
 'BLACKPOOL',
 'BOURNEMOUTH',
 'BRISTOL',
 'CAMPBELTOWN',
 'CARDIFF WALES',
 'CITY OF DERRY (EGLINTON)',
 'DONCASTER SHEFFIELD',
 'DUNDEE',
 'EAST MIDLANDS INTERNATIONAL',
 'EDINBURGH',
 'EXETER',
 'GATWICK',
 'GLASGOW',
 'GLOUCESTERSHIRE',
 'GUERNSEY',
 'HEATHROW',
 'HUMBERSIDE',
 'INVERNESS',
 'ISLAY',
 'ISLE OF MAN',
 'ISLES OF SCILLY (ST.MARYS)',
 'JERSEY',
 'KIRKWALL',
 'LANDS END (ST JUST)',
 'LEEDS BRADFORD',
 'LERWICK (TINGWALL)',
 'LIVERPOOL (JOHN LENNON)',
 'LONDON CITY',
 'LUTON',
 'LYDD',
 'MANCHESTER',
 'NEWCASTLE',
 'NEWQUAY',
 'NORWICH',
 'OXFORD (KIDLINGTON)',
 'PRESTWICK',
 'SCATSTA',
 'SOUTHAMPTON',
 'SOUTHEND',
 'STANSTED',
 'STORNOWAY',
 'SUMBURGH',
 'TEESSIDE INTERNATIONAL AIRPORT',
 'TIREE',
 'WICK JOHN O GROATS']

In [585]:
# Step 4.6:  Check descriptive statistics
pd.set_option('display.float_format', '{:,.0f}'.format)
df_2020_TT_Passengers.describe().T

count      mean        std   min    25%     50%       75%  \
this_period         53     2,020          0 2,020  2,020   2,020     2,020   
last_period         53     2,019          0 2,019  2,019   2,019     2,019   
airport_cluster     53         1          0     1      1       1         1   
rpt_apt_grp_cd      53         3          1     1      3       3         3   
total_pax_tp        53 1,407,712  3,572,683     0 28,048 148,251   908,105   
total_pax_lp        53 5,669,384 13,673,962     0 53,155 803,307 4,675,411   
total_pax_pc        50       -70         14   -87    -79     -75       -65   
term_pax_tp         53 1,406,748  3,572,272     0 28,048 147,921   908,105   
term_pax_lp         53 5,666,629 13,672,890     0 53,155 803,127 4,674,338   
term_pax_pc         50       -71         14   -87    -78     -75       -66   
tran_pax_tp         53       964      1,544     0      0     198     1,399   
tran_pax_lp         53     2,755      7,114     0      0     180     2,016   
tran_pax_pc         31       359      1,551   -99    -73     -48        -9   

                       max  
this_period          2,020  
last_period          2,019  
airport_cluster          2  
rpt_apt_grp_cd           4  
total_pax_tp    22,111,326  
total_pax_lp    80,890,031  
total_pax_pc           -18  
term_pax_tp     22,109,550  
term_pax_lp     80,886,589  
term_pax_pc            -18  
tran_pax_tp          6,539  
tran_pax_lp         33,976  
tran_pax_pc          7,661

***   Major COVID-19 Impact:

-> Passenger traffic dramatically decreased in 2020.


Count	-> 53 (airports	Number of airports reported)

Mean	-> 1,407,712 (passengers	Average traffic per airport during COVID year)

Standard Deviation	->3,572,683	(Large variation, but lower than pre-COVID due to universal drop)

Minimum	-> 0	(Some airports had no activity)

25th Percentile (Q1) ->	28,048	(Small airports experienced extremely low traffic)

Median  -> (Q2)	148,251	(A typical airport handled only ~0.15M passengers)

75th Percentile (Q3) ->	908,105	(Even big airports saw steep decline)

Maximum	-> 22,111,326	(Largest airport traffic fell sharply vs previous year)
_________________________________________________________________________________


     

*** 2020 & 2019 – Total Passengers analysis:

2019 :

Mean Passengers	  -> 5,778,410

 Median Passengers ->	828,992

 Max (busiest airport) -> 80,890,031



2020:

Mean Passengers	 -> 1,407,712

Median Passengers -> 148,251

Max (busiest airport) -> 22,111,326

--------------------------------------------

*Average passengers per airport fell by about 76%

*Typical airport lost ~82% of passengers


*Large airports traffic crashed ~81%









# Data understanding (Table 09 Terminal and Transit Passengers 2021)

In [586]:
# Step 5: Data understanding
# Step 5.1: Understanding the structure of data.
df_2021_TT_Passengers.shape

(53, 16)

In [587]:
# Step 5.2: Display all column names in the dataset
df_2021_TT_Passengers.columns

Index(['rundate', 'this_period', 'last_period', 'airport_cluster',
       'rpt_apt_grp_cd', 'rpt_apt_grp_name', 'rpt_apt_name', 'total_pax_tp',
       'total_pax_lp', 'total_pax_pc', 'term_pax_tp', 'term_pax_lp',
       'term_pax_pc', 'tran_pax_tp', 'tran_pax_lp', 'tran_pax_pc'],
      dtype='object')

In [588]:
# Step 5.3: Preview the first five rows
df_2021_TT_Passengers.head()

rundate  this_period  last_period  airport_cluster  \
0  3/25/2022 10:53:48 AM         2021         2020                1   
1  3/25/2022 10:53:48 AM         2021         2020                1   
2  3/25/2022 10:53:48 AM         2021         2020                1   
3  3/25/2022 10:53:48 AM         2021         2020                1   
4  3/25/2022 10:53:48 AM         2021         2020                1   

   rpt_apt_grp_cd      rpt_apt_grp_name rpt_apt_name  total_pax_tp  \
0               1  London Area Airports      GATWICK       6261814   
1               1  London Area Airports     HEATHROW      19393886   
2               1  London Area Airports  LONDON CITY        720580   
3               1  London Area Airports        LUTON       4674800   
4               1  London Area Airports     SOUTHEND         94367   

   total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  term_pax_pc  \
0      10173431           -38      6260072     10171867          -38   
1      22111326           -12     19392178     22109550          -12   
2        908105           -21       720580       908105          -21   
3       5550821           -16      4673656      5550401          -16   
4        401143           -76        94367       401143          -76   

   tran_pax_tp  tran_pax_lp  tran_pax_pc  
0         1742         1564           11  
1         1708         1776           -4  
2            0            0          NaN  
3         1144          420          172  
4            0            0          NaN

In [589]:
#Step 5.4: Check basic info (data types, missing values, memory use)
df_2021_TT_Passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rundate           53 non-null     object 
 1   this_period       53 non-null     int64  
 2   last_period       53 non-null     int64  
 3   airport_cluster   53 non-null     int64  
 4   rpt_apt_grp_cd    53 non-null     int64  
 5   rpt_apt_grp_name  53 non-null     object 
 6   rpt_apt_name      53 non-null     object 
 7   total_pax_tp      53 non-null     int64  
 8   total_pax_lp      53 non-null     int64  
 9   total_pax_pc      49 non-null     float64
 10  term_pax_tp       53 non-null     int64  
 11  term_pax_lp       53 non-null     int64  
 12  term_pax_pc       48 non-null     float64
 13  tran_pax_tp       53 non-null     int64  
 14  tran_pax_lp       53 non-null     int64  
 15  tran_pax_pc       32 non-null     float64
dtypes: float64(3), int64(10), object(3)
memory usa

In [590]:
# Step 5.5: Display  Airport names  nicely in a list
sorted(df_2021_TT_Passengers['rpt_apt_name'].unique())

['ABERDEEN',
 'ALDERNEY',
 'BARRA',
 'BELFAST CITY (GEORGE BEST)',
 'BELFAST INTERNATIONAL',
 'BENBECULA',
 'BIGGIN HILL',
 'BIRMINGHAM',
 'BLACKPOOL',
 'BOURNEMOUTH',
 'BRISTOL',
 'CAMPBELTOWN',
 'CARDIFF WALES',
 'CITY OF DERRY (EGLINTON)',
 'DONCASTER SHEFFIELD',
 'DUNDEE',
 'EAST MIDLANDS INTERNATIONAL',
 'EDINBURGH',
 'EXETER',
 'GATWICK',
 'GLASGOW',
 'GLOUCESTERSHIRE',
 'GUERNSEY',
 'HEATHROW',
 'HUMBERSIDE',
 'INVERNESS',
 'ISLAY',
 'ISLE OF MAN',
 'ISLES OF SCILLY (ST.MARYS)',
 'JERSEY',
 'KIRKWALL',
 'LANDS END (ST JUST)',
 'LEEDS BRADFORD',
 'LERWICK (TINGWALL)',
 'LIVERPOOL (JOHN LENNON)',
 'LONDON CITY',
 'LUTON',
 'LYDD',
 'MANCHESTER',
 'NEWCASTLE',
 'NEWQUAY',
 'NORWICH',
 'OXFORD (KIDLINGTON)',
 'PRESTWICK',
 'SCATSTA',
 'SOUTHAMPTON',
 'SOUTHEND',
 'STANSTED',
 'STORNOWAY',
 'SUMBURGH',
 'TEESSIDE INTERNATIONAL AIRPORT',
 'TIREE',
 'WICK JOHN O GROATS']

In [591]:
# Step 5.6: Check descriptive statistics
pd.set_option('display.float_format', '{:,.0f}'.format)
df_2021_TT_Passengers.describe().T

count      mean       std   min    25%     50%     75%  \
this_period         53     2,021         0 2,021  2,021   2,021   2,021   
last_period         53     2,020         0 2,020  2,020   2,020   2,020   
airport_cluster     53         1         0     1      1       1       1   
rpt_apt_grp_cd      53         3         1     1      3       3       3   
total_pax_tp        53 1,237,632 3,039,908     0 24,202 126,854 832,529   
total_pax_lp        53 1,407,712 3,572,683     0 28,048 148,251 908,105   
total_pax_pc        49        16        44   -99    -13       8      48   
term_pax_tp         53 1,236,516 3,039,612     0 20,320 126,837 828,224   
term_pax_lp         53 1,406,748 3,572,272     0 28,048 147,921 908,105   
term_pax_pc         48        19        40   -76    -12       8      49   
tran_pax_tp         53     1,117     1,961     0      0     106   1,583   
tran_pax_lp         53       964     1,544     0      0     198   1,399   
tran_pax_pc         32       371     1,754   -98    -29       3      52   

                       max  
this_period          2,021  
last_period          2,020  
airport_cluster          2  
rpt_apt_grp_cd           4  
total_pax_tp    19,393,886  
total_pax_lp    22,111,326  
total_pax_pc           127  
term_pax_tp     19,392,178  
term_pax_lp     22,109,550  
term_pax_pc            127  
tran_pax_tp          9,888  
tran_pax_lp          6,539  
tran_pax_pc          9,902

*** 2021  & 2020– Total Passengers analysis:

2020 :

Mean Passengers	  -> 1,407,712 

 Median Passengers ->	148,251

2021:

Mean Passengers	 -> 1,237,632	

Median Passengers -> 126,854

** Both mean and median dropped.

That means:

2021 had fewer passengers than 2020

→ So there is no increase in 2021

→ Aviation traffic remained very weak and recovery had not started yet

# Data understanding (Table 09 Terminal and Transit Passengers 2022)

In [592]:
# Step 6: Data understanding
# Step 6.1: Understanding the structure of data.
df_2022_TT_Passengers.shape

(51, 16)

In [593]:
# Step 6.2: Display all column names in the dataset
df_2022_TT_Passengers.columns

Index(['rundate', 'this_period', 'last_period', 'airport_cluster',
       'rpt_apt_grp_cd', 'rpt_apt_grp_name', 'rpt_apt_name', 'total_pax_tp',
       'total_pax_lp', 'total_pax_pc', 'term_pax_tp', 'term_pax_lp',
       'term_pax_pc', 'tran_pax_tp', 'tran_pax_lp', 'tran_pax_pc'],
      dtype='object')

In [594]:
# Step 6.3: Preview the first five rows
df_2022_TT_Passengers.head()

rundate  this_period  last_period  airport_cluster  \
0  3/20/2023 10:41:59 AM         2022         2021                1   
1  3/20/2023 10:41:59 AM         2022         2021                1   
2  3/20/2023 10:41:59 AM         2022         2021                1   
3  3/20/2023 10:41:59 AM         2022         2021                1   
4  3/20/2023 10:41:59 AM         2022         2021                1   

   rpt_apt_grp_cd      rpt_apt_grp_name rpt_apt_name  total_pax_tp  \
0               1  London Area Airports      GATWICK      32835381   
1               1  London Area Airports     HEATHROW      61611838   
2               1  London Area Airports  LONDON CITY       3009313   
3               1  London Area Airports        LUTON      13324491   
4               1  London Area Airports     SOUTHEND         89361   

   total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  term_pax_pc  \
0       6261814           424     32831088      6260072          424   
1      19393886           218     61596618     19392178          218   
2        720580           318      3009313       720580          318   
3       4674800           185     13322236      4673656          185   
4         94367            -5        89361        94367           -5   

   tran_pax_tp  tran_pax_lp  tran_pax_pc  
0         4293         1742          146  
1        15220         1708          791  
2            0            0          NaN  
3         2255         1144           97  
4            0            0          NaN

In [595]:
#Step 6.4: Check basic info (data types, missing values, memory use)
df_2022_TT_Passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rundate           51 non-null     object 
 1   this_period       51 non-null     int64  
 2   last_period       51 non-null     int64  
 3   airport_cluster   51 non-null     int64  
 4   rpt_apt_grp_cd    51 non-null     int64  
 5   rpt_apt_grp_name  51 non-null     object 
 6   rpt_apt_name      51 non-null     object 
 7   total_pax_tp      51 non-null     int64  
 8   total_pax_lp      51 non-null     int64  
 9   total_pax_pc      49 non-null     float64
 10  term_pax_tp       51 non-null     int64  
 11  term_pax_lp       51 non-null     int64  
 12  term_pax_pc       48 non-null     float64
 13  tran_pax_tp       51 non-null     int64  
 14  tran_pax_lp       51 non-null     int64  
 15  tran_pax_pc       31 non-null     float64
dtypes: float64(3), int64(10), object(3)
memory usa

In [596]:
# Step 6.5: Display  Airport names  nicely in a list
sorted(df_2022_TT_Passengers['rpt_apt_name'].unique())

['ABERDEEN',
 'ALDERNEY',
 'BARRA',
 'BELFAST CITY (GEORGE BEST)',
 'BELFAST INTERNATIONAL',
 'BENBECULA',
 'BIGGIN HILL',
 'BIRMINGHAM',
 'BLACKPOOL',
 'BOURNEMOUTH',
 'BRISTOL',
 'CAMPBELTOWN',
 'CARDIFF WALES',
 'CITY OF DERRY (EGLINTON)',
 'DONCASTER SHEFFIELD',
 'DUNDEE',
 'EAST MIDLANDS INTERNATIONAL',
 'EDINBURGH',
 'EXETER',
 'GATWICK',
 'GLASGOW',
 'GUERNSEY',
 'HEATHROW',
 'HUMBERSIDE',
 'INVERNESS',
 'ISLAY',
 'ISLE OF MAN',
 'ISLES OF SCILLY (ST.MARYS)',
 'JERSEY',
 'KIRKWALL',
 'LANDS END (ST JUST)',
 'LEEDS BRADFORD',
 'LERWICK (TINGWALL)',
 'LIVERPOOL (JOHN LENNON)',
 'LONDON CITY',
 'LUTON',
 'LYDD',
 'MANCHESTER',
 'NEWCASTLE',
 'NEWQUAY',
 'NORWICH',
 'OXFORD (KIDLINGTON)',
 'PRESTWICK',
 'SOUTHAMPTON',
 'SOUTHEND',
 'STANSTED',
 'STORNOWAY',
 'SUMBURGH',
 'TEESSIDE INTERNATIONAL AIRPORT',
 'TIREE',
 'WICK JOHN O GROATS']

In [597]:
# Step 6.6: Check descriptive statistics
pd.set_option('display.float_format', '{:,.0f}'.format)
df_2022_TT_Passengers.describe().T

count      mean        std   min    25%     50%       75%  \
this_period         51     2,022          0 2,022  2,022   2,022     2,022   
last_period         51     2,021          0 2,021  2,021   2,021     2,021   
airport_cluster     51         1          0     1      1       1         1   
rpt_apt_grp_cd      51         3          1     1      3       3         3   
total_pax_tp        51 4,400,979 10,556,457     0 57,330 445,211 3,237,501   
total_pax_lp        51 1,286,167  3,089,824     0 36,406 127,208   928,730   
total_pax_pc        49       945      5,468   -78     63     140       266   
term_pax_tp         51 4,399,412 10,553,761     0 57,276 444,433 3,237,168   
term_pax_lp         51 1,285,007  3,089,540     0 36,391 127,208   925,382   
term_pax_pc         48       166        133   -78     65     140       236   
tran_pax_tp         51     1,568      4,141     0      0     186     1,291   
tran_pax_lp         51     1,160      1,987     0      0     152     1,646   
tran_pax_pc         31       152        437   -99    -57       2       133   

                       max  
this_period          2,022  
last_period          2,021  
airport_cluster          2  
rpt_apt_grp_cd           4  
total_pax_tp    61,611,838  
total_pax_lp    19,393,886  
total_pax_pc        38,428  
term_pax_tp     61,596,618  
term_pax_lp     19,392,178  
term_pax_pc            599  
tran_pax_tp         24,053  
tran_pax_lp          9,888  
tran_pax_pc          2,078

**Both mean and median increased a LOT, meaning airports finally saw returning passengers.

*** 2022 & 2021 – Total Passengers analysis:

2021:

Mean Passengers	  -> 1,286,167

 Median Passengers ->	127,208	

2022:

Mean Passengers	 -> 4,400,979

Median Passengers -> 445,211

** Both mean and median dropped.

That means:

2022 marks the beginning of recovery after two years of severe decline (2020 & 2021).

Although not fully back to pre-COVID levels, passenger volumes grew rapidly across the UK airport network.

# Data understanding (Table 09 Terminal and Transit Passengers 2023)

In [598]:
# Step 7: Data understanding
# Step 7.1: Understanding the structure of data.
df_2023_TT_Passengers.shape

(52, 16)

In [599]:
# Step 7.2: Display all column names in the dataset
df_2023_TT_Passengers.columns

Index(['rundate', 'this_period', 'last_period', 'airport_cluster',
       'rpt_apt_grp_cd', 'rpt_apt_grp_name', 'rpt_apt_name', 'total_pax_tp',
       'total_pax_lp', 'total_pax_pc', 'term_pax_tp', 'term_pax_lp',
       'term_pax_pc', 'tran_pax_tp', 'tran_pax_lp', 'tran_pax_pc'],
      dtype='object')

In [600]:
# Step 7.3: Preview the first five rows
df_2023_TT_Passengers.head()

rundate  this_period  last_period  airport_cluster  \
0  3/19/2024 1:15:09 PM         2023         2022                1   
1  3/19/2024 1:15:09 PM         2023         2022                1   
2  3/19/2024 1:15:09 PM         2023         2022                1   
3  3/19/2024 1:15:09 PM         2023         2022                1   
4  3/19/2024 1:15:09 PM         2023         2022                1   

   rpt_apt_grp_cd      rpt_apt_grp_name rpt_apt_name  total_pax_tp  \
0               1  London Area Airports      GATWICK      40897656   
1               1  London Area Airports     HEATHROW      79180434   
2               1  London Area Airports  LONDON CITY       3429684   
3               1  London Area Airports        LUTON      16403784   
4               1  London Area Airports     SOUTHEND        146072   

   total_pax_lp  total_pax_pc  term_pax_tp  term_pax_lp  term_pax_pc  \
0      32835381            25     40894242     32831088           25   
1      61611838            29     79149042     61596618           28   
2       3009313            14      3429684      3009313           14   
3      13324491            23     16399866     13322236           23   
4         89361            63       146072        89361           63   

   tran_pax_tp  tran_pax_lp  tran_pax_pc  
0         3414         4293          -20  
1        31392        15220          106  
2            0            0          NaN  
3         3918         2255           74  
4            0            0          NaN

In [601]:
#Step 7.4: Check basic info (data types, missing values, memory use)
df_2023_TT_Passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   rundate           52 non-null     object 
 1   this_period       52 non-null     int64  
 2   last_period       52 non-null     int64  
 3   airport_cluster   52 non-null     int64  
 4   rpt_apt_grp_cd    52 non-null     int64  
 5   rpt_apt_grp_name  52 non-null     object 
 6   rpt_apt_name      52 non-null     object 
 7   total_pax_tp      52 non-null     int64  
 8   total_pax_lp      52 non-null     int64  
 9   total_pax_pc      49 non-null     float64
 10  term_pax_tp       52 non-null     int64  
 11  term_pax_lp       52 non-null     int64  
 12  term_pax_pc       49 non-null     float64
 13  tran_pax_tp       52 non-null     int64  
 14  tran_pax_lp       52 non-null     int64  
 15  tran_pax_pc       29 non-null     float64
dtypes: float64(3), int64(10), object(3)
memory usa

In [602]:
# Step 7.5: Display  Airport names  nicely in a list
sorted(df_2023_TT_Passengers['rpt_apt_name'].unique())

['ABERDEEN',
 'ALDERNEY',
 'BARRA',
 'BELFAST CITY (GEORGE BEST)',
 'BELFAST INTERNATIONAL',
 'BENBECULA',
 'BIGGIN HILL',
 'BIRMINGHAM',
 'BLACKPOOL',
 'BOURNEMOUTH',
 'BRISTOL',
 'CAMPBELTOWN',
 'CARDIFF WALES',
 'CITY OF DERRY (EGLINTON)',
 'DONCASTER SHEFFIELD',
 'DUNDEE',
 'EAST MIDLANDS INTERNATIONAL',
 'EDINBURGH',
 'EXETER',
 'FARNBOROUGH',
 'GATWICK',
 'GLASGOW',
 'GUERNSEY',
 'HEATHROW',
 'HUMBERSIDE',
 'INVERNESS',
 'ISLAY',
 'ISLE OF MAN',
 'ISLES OF SCILLY (ST.MARYS)',
 'JERSEY',
 'KIRKWALL',
 'LANDS END (ST JUST)',
 'LEEDS BRADFORD',
 'LERWICK (TINGWALL)',
 'LIVERPOOL (JOHN LENNON)',
 'LONDON CITY',
 'LUTON',
 'LYDD',
 'MANCHESTER',
 'NEWCASTLE',
 'NEWQUAY',
 'NORWICH',
 'OXFORD (KIDLINGTON)',
 'PRESTWICK',
 'SOUTHAMPTON',
 'SOUTHEND',
 'STANSTED',
 'STORNOWAY',
 'SUMBURGH',
 'TEESSIDE INTERNATIONAL AIRPORT',
 'TIREE',
 'WICK JOHN O GROATS']

In [603]:
# Step 7.6: Check descriptive statistics
pd.set_option('display.float_format', '{:,.0f}'.format)
df_2023_TT_Passengers.describe().T

count      mean        std   min    25%     50%       75%  \
this_period         52     2,023          0 2,023  2,023   2,023     2,023   
last_period         52     2,022          0 2,022  2,022   2,022     2,022   
airport_cluster     52         1          0     1      1       1         1   
rpt_apt_grp_cd      52         3          1     1      3       3         3   
total_pax_tp        52 5,303,265 13,244,849     0 38,747 420,968 3,946,724   
total_pax_lp        52 4,316,827 10,470,501     0 47,806 413,717 3,211,934   
total_pax_pc        49        16         25   -98      6      16        25   
term_pax_tp         52 5,301,559 13,240,630     0 37,336 420,964 3,946,724   
term_pax_lp         52 4,315,290 10,467,823     0 47,414 413,316 3,211,767   
term_pax_pc         49        16         25   -98      6      17        25   
tran_pax_tp         52     1,706      5,168     0      0      88       986   
tran_pax_lp         52     1,538      4,106     0      0     186     1,250   
tran_pax_pc         29       149        621   -93    -45     -19        74   

                       max  
this_period          2,023  
last_period          2,022  
airport_cluster          2  
rpt_apt_grp_cd           4  
total_pax_tp    79,180,434  
total_pax_lp    61,611,838  
total_pax_pc            67  
term_pax_tp     79,149,042  
term_pax_lp     61,596,618  
term_pax_pc             67  
tran_pax_tp         31,392  
tran_pax_lp         24,053  
tran_pax_pc          3,233

***UK airports overall recovered strongly in 2023.

*** 2022 & 2023 – Total Passengers analysis:

2022:

Mean Passengers	  -> 4,400,979

 Median Passengers ->	445,211

2023:

Mean Passengers	 -> 5,303,265

Median Passengers -> 420,968

*Mean increased

*Median slightly decreased


That means: 

On average, airports saw better recovery (more passengers than 2022)

 -> But typical/medium-sized airports didn’t grow as much

 -> Large airports (e.g., Heathrow) drove the improvement

🇬🇧 Aviation recovery strengthened in 2023

# Data cleaning 

In [604]:
# Step 8: Data cleaning.
# Step 8.1: Define airports of interest (Scotland + comparison)

airports_scotland = [
    "ABERDEEN",
    "EDINBURGH",
    "GLASGOW",
    "INVERNESS",
    "PRESTWICK"
]

airports_compare = [
    "HEATHROW",
    "GATWICK",
    "MANCHESTER"
]

airports_of_interest = airports_scotland + airports_compare
airports_of_interest

['ABERDEEN',
 'EDINBURGH',
 'GLASGOW',
 'INVERNESS',
 'PRESTWICK',
 'HEATHROW',
 'GATWICK',
 'MANCHESTER']

In [605]:
# Step 8.2: Tell Python where each file is
file_paths = {
    2019: "2019/Table 09 Terminal and Transit Passengers2019.csv",
    2020: "2020/Table 09 Terminal and Transit Passengers2020.csv",
    2021: "2021/Table_09_Terminal_and_Transit_Passengers2021.csv",
    2022: "2022/Table_09_Terminal_and_Transit_Passengers2022.csv",
    2023: "2023/Table_09_Terminal_and_Transit_Passengers_2023.csv",
}

In [606]:
# Step 8.3: Function to Load and Clean Passenger Data (Table 09)


import pandas as pd

def load_clean_passengers(year):
    """
    Load and clean Table 09 (Terminal & Transit Passengers) for one year.
    Uses common columns: this_period, rpt_apt_name, total_pax_tp.
    Returns: Year, Airport, Total_Passengers
    """
    path = file_paths[year]          # pick the right file
    df = pd.read_csv(path)

    # 1. keep only the columns we need
    cols_keep = ["this_period", "rpt_apt_name", "total_pax_tp"]
    df_clean = df[cols_keep].copy()

    # 2. Year from this_period (works for 2019 or 201901 etc.)
    df_clean["Year"] = df_clean["this_period"].astype(str).str[:4].astype(int)

    # 3. clean airport names
    df_clean["Airport"] = df_clean["rpt_apt_name"].str.upper().str.strip()

    # 4. keep only selected airports
    df_clean = df_clean[df_clean["Airport"].isin(airports_of_interest)]

    # 5. final tidy columns
    df_clean = df_clean[["Year", "Airport", "total_pax_tp"]].rename(
        columns={"total_pax_tp": "Total_Passengers"}
    )

    return df_clean



***What this function does:***

->Loads the correct CSV file based on the year

->Keeps only useful columns

->Extracts the proper Year value

->Standardizes airport names (uppercase + remove spaces)

->Filters to only key airports we want to analyze

->Renames columns to a clean and simple format

->Returns a tidy dataframe ready for combining

Output Columns:

Year

Airport

Total_Passengers

In this step, we create a ***reusable*** function to clean the Terminal & Transit passenger dataset for any selected year.



***-> This doesn’t change your CSV files on disk – it only creates a cleaned dataframe in memory.

In [607]:
# Step 8.4: Automatically clean and combine 2019–2023
years = [2019, 2020, 2021, 2022, 2023]

cleaned_list = []        # A list to store cleaned datasets

for y in years:          # Loop to clean each year
    df_year = load_clean_passengers(y)
    print(f"{y}: {df_year.shape}")   # quick check
    cleaned_list.append(df_year)

df_passengers = pd.concat(cleaned_list, ignore_index=True)  # Combine all years together

print("Combined dataset:", df_passengers.shape)
df_passengers.head()


2019: (8, 3)
2020: (8, 3)
2021: (8, 3)
2022: (8, 3)
2023: (8, 3)
Combined dataset: (40, 3)


Year    Airport  Total_Passengers
0  2019    GATWICK          46576473
1  2019   HEATHROW          80890031
2  2019   ABERDEEN           2912883
3  2019  EDINBURGH          14737497
4  2019    GLASGOW           8847100

In [ ]:
# Step 8.5: Save cleaned final dataset for dashboard visualization
df_passengers.to_csv("cleaned_data.csv", index=False)


***This will create a file: cleaned_data.csv

#  Data Visualizations 

In [608]:
# Step 9 COVID impact: total passengers by year

year_totals = (
    df_passengers
    .groupby("Year", as_index=False)["Total_Passengers"]
    .sum()
)

year_totals

Year  Total_Passengers
0  2019         184939628
1  2020          46067000
2  2021          38349601
3  2022         138685016
4  2023         173487631

In [609]:
# Step 10: Visualizing COVID Impact: Total Passengers by Year (Bar Chart)
import plotly.express as px

fig_year = px.bar(
    year_totals,
    x="Year",
    y="Total_Passengers",
    title="Total Passengers (Selected Airports) – 2019–2023",
    text_auto=True
)
fig_year.show()


nsights from Passenger Trends (2019–2023)

1️⃣ 2019 – Baseline / Normal Year

Passenger traffic was at its highest (≈ 185 Million).

Represents normal pre-COVID aviation demand.

2️⃣ 2020 – Severe COVID Shock

Traffic collapsed to ≈ 46M (⬇ ~75% drop from 2019).

Lockdowns, border closures, and travel restrictions halted air travel.

3️⃣ 2021 – Slow & Uncertain Recovery

Slight decrease vs 2020 (≈ 38M).

Continued travel restrictions & variant waves (Delta, Omicron).

Aviation recovery remained weak.

4️⃣ 2022 – Strong Recovery Begins

Traffic jumped to ≈ 139M (⬆ >250% from 2021).

International routes reopening, demand returning.

People resumed travel for work, family, tourism.

5️⃣ 2023 – Near Full Recovery

Reaches ≈ 173M, close to 2019 levels (only ~6% behind).

Confidence in travel restored

Airlines improved schedules and capacity

Heathrow-	Highest passengers overall. Biggest fall in 2020, strongest recovery by 2023.

Gatwick-	Large drop during pandemic but bounced back close to 2019 levels.

Manchester-	Solid recovery after 2021, continues growing toward pre-COVID levels.

Edinburgh-Clear recovery trend; close to returning to 2019 traffic by 2023.

Glasgow	-Recovered slower than Edinburgh but still improving year-over-year.

Aberdeen-	Smaller decline compared to others due to essential travel demand.

Inverness & Prestwick	-Low passenger volumes and slower growth—still below 2019 levels.

In [610]:

# Step 11: Passenger trends by airport (colorful version)

# Sort for better line shaping
df_trend = df_passengers.sort_values(["Airport", "Year"])

# Apply a colorful theme
px.defaults.template = "plotly"  # modern colorful theme

fig_trend = px.line(
    df_trend,
    x="Year",
    y="Total_Passengers",
    color="Airport",
    markers=True,
    title="Passenger Trends by Airport (2019–2023)",
    line_group="Airport"
)

fig_trend.update_layout(
    title_font=dict(size=20),
    xaxis_title="Year",
    yaxis_title="Total Passengers",
    legend_title="Airport",
    hovermode="x unified",
    
)

fig_trend.update_traces(mode="lines+markers")  # ensures markers appear
fig_trend.show()


Heathrow	- Huge decline in 2020–21, strong rebound by 2023	Largest UK airport — 2023 traffic almost back to 2019 level

Gatwick	-Very deep COVID hit — slower recovery	Heavy dependence on international leisure travel caused bigger drop

Manchester	-Sharp fall in 2020–21, good growth in 2022–23	Strong comeback but still slightly below 2019 peak

Edinburgh	-Scotland’s busiest — V-shaped recovery	Traffic keeps rising but not yet fully recovered to 2019

Glasgow	-Moderate rebound after steep drop	Still significantly below pre-pandemic demand in 2023

Aberdeen-	Less impacted — business travel helped stabilize	Gradual recovery but remains lower than pre-COVID

Inverness	-Small airport — slow recovery	2023 still far below 2019 levels due to limited routes

Prestwick	- Minimal passengers compared to others.Very small operation, weak recovery — lowest among all airports

In [611]:
# step 12: Recovery table: 2019 vs 2023 for each airport

 #Select only 2019 data
base_2019 = (
    df_passengers[df_passengers["Year"] == 2019]
    [["Airport", "Total_Passengers"]]
    .rename(columns={"Total_Passengers": "Pax_2019"})
)

#Select only 2023 data
base_2023 = (
    df_passengers[df_passengers["Year"] == 2023]
    [["Airport", "Total_Passengers"]]
    .rename(columns={"Total_Passengers": "Pax_2023"})
)
# Join 2019 + 2023 Data
recovery = base_2019.merge(base_2023, on="Airport", how="left")

# Calculate Recovery Percentage

recovery["Recovery_%"] = (recovery["Pax_2023"] / recovery["Pax_2019"]) * 100

#Sort Best → Worst Recovery
recovery.sort_values("Recovery_%", ascending=False)



Airport  Pax_2019  Pax_2023  Recovery_%
1    HEATHROW  80890031  79180434          98
3   EDINBURGH  14737497  14396794          98
6  MANCHESTER  29397357  28096783          96
0     GATWICK  46576473  40897656          88
5   INVERNESS    938232    802338          86
4     GLASGOW   8847100   7358828          83
7   PRESTWICK    640055    524880          82
2    ABERDEEN   2912883   2229918          77

UK:

Heathrow nearly back to 2019 levels

Gatwick & Manchester showing strong upward trend

Scotland:

Edinburgh leads recovery in Scotland (best performance)

Glasgow, Aberdeen still far from pre-pandemic traffic
________________________________________________________
Aviation recovery is not equal across the UK.

Large English airports recovered much faster than smaller Scottish airports.

In [612]:
# Step 13: Add Region column (Scotland / England)

region_map = {
    "ABERDEEN": "Scotland",
    "EDINBURGH": "Scotland",
    "GLASGOW": "Scotland",
    "INVERNESS": "Scotland",
    "PRESTWICK": "Scotland",
    "HEATHROW": "England",
    "GATWICK": "England",
    "MANCHESTER": "England"
}

df_passengers["Region"] = df_passengers["Airport"].map(region_map)

df_passengers.head()


Year    Airport  Total_Passengers    Region
0  2019    GATWICK          46576473   England
1  2019   HEATHROW          80890031   England
2  2019   ABERDEEN           2912883  Scotland
3  2019  EDINBURGH          14737497  Scotland
4  2019    GLASGOW           8847100  Scotland